### ISSUE TITLE
BUG: Function parameter types are not checked inside other function bodies 

### ISSUE THREAD

**Issue by lingz** [python/mypy#5689]
Simple repro using mypy 0.630:

`test.py`
```python
def test(x: int):
    print(x)

test('abc')

def test2():
    test('abc')
```

`mypy test.py`
Gives: `test.py:4: error: Argument 1 to "test" has incompatible type "str"; expected "int"`

Clearly it should also report the error in `test2`, but it seems parameters are not checked inside function bodies? In fact the following file has no type errors:

`test.py`
```python
def test(x: int):
    print(x)

def test2():
    test('abc')
```

**Comment by gvanrossum** [comment]
This is because `test2()` is not annotated. Mypy doesn't check inside unannotated functions. If you add `-> None` to `test2()` it will detect the error.

Good luck!

**Comment by lingz** [comment]
Thanks for the quick reply, I just read the docs and saw this, though I wonder if there is a reasoning for this as it seems counterintuitive as:

- mypy has a case for type inference, so it does not feel necessary to annotate every function definition, as these feel like they would be inferred
- there are many real use cases where the return type is just noise to annotate, for example, I was wondering why none of my unit tests have type checking turned on.

In fact not knowing this behavior made us spend a lot of time thinking we had setup our build toolchain was wrong, because type checking sporadically worked / stopped working but it seems that this was the root cause.

**Comment by Michael0x2a** [comment]
AFAIK, the reason for this behavior is to help make sure mypy can integrate seamlessly with existing Python code. Suddenly trying to type-check everything can often lead to a lot of errors, and for large enough codebases, it isn't really feasible to try and fix all those errors at once. So, it's useful to have these sorts of mechanisms that let you distinguish between code that's ready to be type-checked vs code that isn't.

In any case, if you want to tune this behavior, you can configure how mypy handles untyped definitions and calls via [command line flags](https://mypy.readthedocs.io/en/latest/command_line.html#untyped-definitions-and-calls) or a [config file](https://mypy.readthedocs.io/en/latest/config_file.html#untyped-definitions-and-calls).

For example, I think either the `--disallow-untyped-defs` or the `--check-untyped-defs` flag would be useful here. (Note: the latter flag will just make mypy assume any unannotated functions accept and return `Any` (the dynamic type) and typecheck accordingly. Mypy can only do local type inference, so won't be able to use the global context of your program to infer the signature of unannotated functions.)

Maybe the correct thing to do here is to add some more emphasis about this to the "getting started" page of the docs. That page mentions this behavior, but it's somewhat easy to miss imo.

**Comment by lingz** [comment]
I think more emphasis is a good approach, or it seems like `--check-untyped-defs` could be the default, and we could have `--skip-untyped-defs`. As one would assume someone migrating a large codebase would be more sophisticated and likely is not using defaults anyway. Though this is definitely one of the more "unintuitive" parts of mypy. 

To draw a parallel: typescript does exactly that, defaults to type checking everything, but annotating all unannotated (and non-inferrable returns) with Any type. They also have the idea that you can gradually migrate a large code base and it feels much more intuitive to me.
